content/docs/guides/testing/unit.md

```diff
@@ -0,0 +1,964 @@
+        UserData:       pulumi.String(userData),        // start a simple web server
+    })
+    if err != nil {
+        return nil, err
+    }
+
+    return &infrastructure{
+        group:  group,
+        server: server,
+    }, nil
+}
+```
+
+{{% /choosable %}}
+{{% choosable language "csharp" %}}
+
+**WebserverStack.cs**:
+
+``` csharp
+using Pulumi;
+using Pulumi.Aws.Ec2;
+using Pulumi.Aws.Ec2.Inputs;
+
+public class WebserverStack : Stack
+{
+    public WebserverStack()
+    {
+        var group = new SecurityGroup("web-secgrp", new SecurityGroupArgs
+        {
+            Ingress =
+            {
+                new SecurityGroupIngressArgs { Protocol = "tcp", FromPort = 22, ToPort = 22, CidrBlocks = { "0.0.0.0/0" } },
+                new SecurityGroupIngressArgs { Protocol = "tcp", FromPort = 80, ToPort = 80, CidrBlocks = { "0.0.0.0/0" } }
+            }
+        });
+
+        var userData = "#!/bin/bash echo \"Hello, World!\" > index.html nohup python -m SimpleHTTPServer 80 &";
+
+        var server = new Instance("web-server-www", new InstanceArgs
+        {
+            InstanceType = "t2.micro",
+            SecurityGroups = { group.Name }, // reference the group object above
+            Ami = "ami-c55673a0",            // AMI for us-east-2 (Ohio)
+            UserData = userData              // start a simple webserver
+        });
+    }
+}
+```
+
+{{% /choosable %}}
+
+This basic Pulumi program allocates a security group and an instance. Notice, however, that we are violating all three of the rules stated above&mdash;let's write some tests!
+
+## Install the unit testing framework
+
+You are free to use your favorite frameworks and libraries for writing unit tests and assertions.
+
+{{% choosable language "typescript" %}}
+
+This guide uses Mocha as the testing framework. [Install Mocha](https://mochajs.org/#installation) to your development environment.
+
+Then, install additional NPM modules to your program:
+
+```bash
+npm install mocha @types/mocha ts-node
+```
+
+{{% /choosable %}}
+
+{{% choosable language python %}}
+
+We use the built-in [`unittest`](https://docs.python.org/3/library/unittest.html) framework, so no need to install anything.
+
+{{% /choosable %}}
+
+{{% choosable language go %}}
+
+We use the built-in `go test` command, so no need to install anything.
+
+{{% /choosable %}}
+
+{{% choosable language "csharp" %}}
+
+We use [NUnit](https://nunit.org/) test framework to define and run the tests, [Moq](https://github.com/moq/moq4) for mocks, and [FluentAssertions](https://github.com/fluentassertions/fluentassertions) for assertions.
+
+Install the corresponding NuGet packages to your program:
+
+```bash
+dotnet add package NUnit
+dotnet add package NUnit3TestAdapter
+dotnet add package Moq
+dotnet add package FluentAssertions
+```
+
+{{% /choosable %}}
+
+## Add Mocks
+
+Let's add the following code to mock the external calls to the Pulumi CLI.
+
```

Comment:
I think it would be good to add a pointer here to https://www.pulumi.com/docs/reference/pkg/nodejs/pulumi/pulumi/runtime/#Mocks so that end users know where these come from and what the interface looks like.

Reply:
Added `The definition of the mocks interface is available at the runtime API reference page.` down below